src/views/Users/ViewUser/index.jsx

```diff
@@ -0,0 +1,470 @@
+import PlusIcon from 'mdi-react/PlusIcon';
+import ErrorPanel from '@mozilla-frontend-infra/components/ErrorPanel';
+import Spinner from '@mozilla-frontend-infra/components/Spinner';
+import Dashboard from '../../../components/Dashboard';
+import SpeedDial from '../../../components/SpeedDial';
+import useAction from '../../../hooks/useAction';
+import { getUserInfo } from '../../../utils/Users';
+import zip from '../../../utils/zip';
+
+const useStyles = makeStyles(theme => ({
+  fab: {
+    ...theme.mixins.fab,
+    right: theme.spacing(12),
+  },
+  fullWidth: {
+    width: '100%',
+  },
+  addGrid: {
+    marginTop: theme.spacing(0),
+  },
+  gridWithIcon: {
+    marginTop: theme.spacing(3),
+  },
+}));
+
+function ViewUser({ isNewUser, ...props }) {
+  const getEmptyPermission = () => ({
+    name: '',
+    options: {
+      products: [],
+      actions: [],
+    },
+    metadata: {
+      isAdditional: true,
+    },
+  });
+  const {
+    match: {
+      params: { username: existingUsername },
+    },
+  } = props;
+  const classes = useStyles();
+  const [username, setUsername] = useState('');
+  const [roles, setRoles] = useState([]);
+  const [originalRoles, setOriginalRoles] = useState([]);
+  const [additionalRoles, setAdditionalRoles] = useState([]);
+  const [permissions, setPermissions] = useState([]);
+  const [originalPermissions, setOriginalPermissions] = useState([]);
+  const [additionalPermissions, setAdditionalPermissions] = useState(
+    isNewUser ? [getEmptyPermission()] : []
+  );
+  const [userAction, fetchUser] = useAction(getUserInfo);
+  const [saveAction, saveUser] = useAction(() => {});
+  const isLoading = userAction.loading;
+  const error = userAction.error || saveAction.error;
+
+  useEffect(() => {
+    if (!isNewUser) {
+      fetchUser(existingUsername).then(result => {
+        const roles = Object.keys(result.data.data.roles).map(name => ({
+          name,
+          data_version: result.data.data.roles[name].data_version,
+        }));
+        const permissions = Object.keys(result.data.data.permissions).map(
+          name => {
+            const details = result.data.data.permissions[name];
+
+            return {
+              name,
+              options: details.options,
+              data_version: details.data_version,
+              metadata: {
+                isAdditional: false,
+              },
+            };
+          }
+        );
+
+        setUsername(result.data.data.username);
+        setRoles(roles);
+        setOriginalRoles(JSON.parse(JSON.stringify(roles)));
+        setPermissions(permissions);
+        setOriginalPermissions(JSON.parse(JSON.stringify(permissions)));
+      });
+    }
+  }, []);
+
+  const handleUsernameChange = ({ target: { value } }) => setUsername(value);
+  // this currently throws an error about controlled components
+  // (https://reactjs.org/docs/forms.html#controlled-components)
+  // not sure how to fix it yet
```

Comment:
I suspect this is because `role.name` is `undefined`when a user clicks the "+" button. Initially when a user clicks the "+" button, the value prop is set to `undefined` (uncontrolled). Typing in it would make the value prop no longer undefined (controlled) which explains why `A component is changing an uncontrolled input of type text to be controlled` shows up in the console. In order to remove the error, the `TextField` component should be defined (controlled) from the start.

Reply:
Thank you!!